File: 3932-count-k-th-roots-in-a-range/3932-count-k-th-roots-in-a-range.erl
-spec count_kth_roots(L :: integer(), R :: integer(), K :: integer()) -> integer().
count_kth_roots(L, R, K) ->
    case K of
        1 -> R - L + 1;
        _ -> iterate_roots(0, 0, L, R, K)
    end.

iterate_roots(X, Count, L, R, K) ->
    Value = bounded_power(X, K, R),
    case Value > R of
        true -> Count;
        false ->
            case Value >= L of
                true -> iterate_roots(X + 1, Count + 1, L, R, K);
                false -> iterate_roots(X + 1, Count, L, R, K)
            end
    end.

bounded_power(_, 0, _) ->
    1;
bounded_power(X, K, R) ->
    Partial = bounded_power(X, K - 1, R),
    Value = X * Partial,
    case Value > R of
        true -> R + 1;
        false -> Value
    end.

File: 3932-count-k-th-roots-in-a-range/3932-count-k-th-roots-in-a-range.ex
defmodule Solution do
  @spec count_kth_roots(l :: integer, r :: integer, k :: integer) :: integer
  def count_kth_roots(l, r, k) do
    if k == 1 do
      r - l + 1
    else
      iterate(0, 0, l, r, k)
    end
  end

  defp iterate(x, count, l, r, k) do
    value = power_with_limit(x, k, r)

    cond do
      value > r -> count
      value >= l -> iterate(x + 1, count + 1, l, r, k)
      true -> iterate(x + 1, count, l, r, k)
    end
  end

  defp power_with_limit(_, 0, _) do
    1
  end

  defp power_with_limit(x, k, r) do
    value = x * power_with_limit(x, k - 1, r)
    if value > r, do: r + 1, else: value
  end
end